mitiq/folding_cirq.py

```diff
@@ -0,0 +1,61 @@
+import random
+import cirq
+
+def local_folding(circuit, stretch, sampling=False):
+    """Applies the map G -> G G^dag G to ((stretch-1)*len(circuit)//2) to a subset of gates 
+    of the input circuit. For large stretch factors the function is recursively repeated. 
+    Returns a circuit of depth approximately equal to stretch*len(circuit).
+    The stretch factor can be any real number >= 1."""
+
+    if not (stretch >= 1):
+        raise ValueError("The stretch factor must be a real number >= 1.")
+
+    out = cirq.Circuit()
+
+    if stretch <= 3:
+        # select a fraction of subindices
+        d = len(circuit)
+        f = int(d * (stretch - 1) / 2)
+        if sampling == True:
```

Comment:
The boolean sampling kwarg indicates that there is a modularity to this function. This function could be refactored a bit to take advantage of this modularity and make it more extensible. For example:

```
def local_folding(circuit, stretch, stretcher):
    if not (stretch >= 1):
        raise ValueError("The stretch factor must be a real number >= 1.")
    if stretch <= 3:
        return stretcher(circuit)
    else:
        # recursive application for large stretching
        _ = local_folding(circuit, 3, sampling)

        return local_folding(_, stretch / 3, sampling)

def sampling_stretcher(circuit):
        d = len(circuit)
        f = int(d * (stretch - 1) / 2)
        # indices of f random gates
        sub_indices = random.sample(range(d), f)
        return append_folds(circuit, sub_indicies)

def start_stretcher(circuit):
        d = len(circuit)
        f = int(d * (stretch - 1) / 2)
        # indices of the first f gates
        sub_indices = list(range(f))
        return append_folds(circuit, sub_indicies)

def append_folds(circuit, sub_indicies):
        out = cirq.Circuit()
        # sequentially append gates to out, folding only if j is in sub_indices
        for j, gate in enumerate(circuit):
            out += gate
            if j in sub_indices:
                out += cirq.inverse(gate)
                out += gate
         return out
```

Note this may have some typos as I haven't tested it out. Written in this manner makes it easier to add different ways of selecting the indicies or different ways of inserting / appending the gates.